### PR TITLE
[INLONG-10286][Agent] Update the MQTT Source

### LINKED ISSUE #10286: [Improve][Agent] Update the MQTT Source
### Description

the latest Agent plugin factored,  MQTT Source needs to be updated, you can refer to [the development guide](https://inlong.apache.org/zh-CN/docs/next/design_and_concept/how_to_write_plugin_agent) and [MongoDB source](https://github.com/apache/inlong/pull/10007).

image

Requirements:
1. Optimize the logic and performance
2. A set of UTs

**This issue is a dedicated issue for the 2024 Tencent Rhino-bird Open Source Talent Training event and is only available for college students to claim**
【Claim the issue】Sign up for the corresponding issue in columns M~R of the [Research and Study Base](https://docs.qq.com/sheet/DTk9YV3pXSklZZ29D) "issue Camp" to be considered as **successfully claimed**.
【Complete the issue】For students who have claimed the issue, please also reply "Successfully claimed this issue" in the comment section of this issue; if there is no progress submitted within 7 days, including but not limited to comments, commits, or Pull Requests, it will be considered as voluntarily giving up the issue and the committee will release the issue to the next person in line.

### InLong Component

InLong Agent

### Are you willing to submit PR?

- [ ] Yes, I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qy-liuhuo**
Fixes #10286 

### Motivation

update MQTT data source for Agent

### Modifications

1. Update MQTT Task、Instance and source
2. Update TaskConstants

### Verifying this change

image

### CODE REVIEW

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java [hunk 1]
```diff
@@ -20,63 +20,57 @@
...
-        }
-        if (CollectionUtils.isNotEmpty(readerList)) {
-            sourceMetric.sourceSuccessCount.incrementAndGet();
-        } else {
-            sourceMetric.sourceFailCount.incrementAndGet();
-        }
+        LOGGER.info("splited!!!");
+        String topics = conf.get(TaskConstants.TASK_MQTT_TOPIC, StringUtils.EMPTY);
+        List<Reader> readerList = splitSqlJob(topics, instanceId);
+        // if (CollectionUtils.isNotEmpty(readerList)) {
```
Comment (vernedeng) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java:69]: please remove these code

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java [hunk 2]
```diff
@@ -20,63 +20,57 @@
...
-        String topics = conf.get(JOB_MQTTJOB_TOPICS, StringUtils.EMPTY);
-        List<Reader> readerList = null;
-        if (StringUtils.isNotEmpty(topics)) {
-        }
-        if (CollectionUtils.isNotEmpty(readerList)) {
-            sourceMetric.sourceSuccessCount.incrementAndGet();
-        } else {
-            sourceMetric.sourceFailCount.incrementAndGet();
-        }
+        LOGGER.info("splited!!!");
```
Comment (vernedeng) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java:66]: please log more specific information like "start to split mqtt source, conf={}"

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java [hunk 3]
```diff
@@ -87,27 +81,51 @@ protected String getThreadName() {
...
     @Override
     protected void initSource(InstanceProfile profile) {
-
+        try {
+            LOGGER.info("MqttSource init: {}", profile.toJsonStr());
+            topic = profile.get(TaskConstants.TASK_MQTT_TOPIC);
+            mqttReader = new MqttReader(topic);
+            mqttReader.init(profile);
+        } catch (Exception e) {
+            stopRunning();
```
Comment (vernedeng) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java:90]: please log the exception why stop initing source

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java [hunk 4]
```diff
@@ -246,4 +237,26 @@ public void destroy() {
             }
         }
     }
+
+    public static void main(String[] args) throws InterruptedException {
```
Comment (vernedeng) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java:241]: please remove test code

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/task/MqttTask.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        if (!profile.hasKey(TaskConstants.TASK_MQTT_PASSWORD)) {
+            LOGGER.info("task profile needs password");
+            return false;
+        }
+        return true;
+    }
+
+    @Override
+    protected int getInstanceLimit() {
+        return DEFAULT_INSTANCE_LIMIT;
```
Comment (vernedeng) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/task/MqttTask.java:73]: please make instanceLimit configurable

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/task/MqttTask.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class MqttTask extends AbstractTask {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MqttTask.class);
+
+    private String topic;
+
+    private boolean isAdded = false;
```
Comment (vernedeng) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/task/MqttTask.java:40]: please use AtiomicBoolean to ensure thread safety

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java [hunk 7]
```diff
@@ -20,63 +20,57 @@
...
-        if (StringUtils.isNotEmpty(topics)) {
-        }
-        if (CollectionUtils.isNotEmpty(readerList)) {
-            sourceMetric.sourceSuccessCount.incrementAndGet();
-        } else {
-            sourceMetric.sourceFailCount.incrementAndGet();
-        }
+        LOGGER.info("splited!!!");
+        String topics = conf.get(TaskConstants.TASK_MQTT_TOPIC, StringUtils.EMPTY);
+        List<Reader> readerList = splitSqlJob(topics, instanceId);
```
Comment (aloyszhang) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java:68]: Local variable 'readerList' is redundant, just return `splitSqlJob(topics, instanceId)` is ok.

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java [hunk 8]
```diff
@@ -88,22 +77,22 @@ public MqttReader(String topic) {
...
-        cleanSession = jobConf.getBoolean(JOB_MQTT_CLEAN_SESSION, false);
-        automaticReconnect = jobConf.getBoolean(JOB_MQTT_AUTOMATIC_RECONNECT, true);
-        qos = jobConf.getInt(JOB_MQTT_QOS, 1);
-        mqttVersion = jobConf.getInt(JOB_MQTT_VERSION, MqttConnectOptions.MQTT_VERSION_DEFAULT);
-
+        userName = jobConf.get(TaskConstants.TASK_MQTT_USERNAME);
+        password = jobConf.get(TaskConstants.TASK_MQTT_PASSWORD);
+        serverURI = jobConf.get(TaskConstants.TASK_MQTT_SERVER_URI);
+
+        // topic = jobConf.get(TaskConstants.TASK_MQTT_TOPIC);
```
Comment (aloyszhang) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java:86]: remove this line if not used.

Comment (aloyszhang) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java:86]: remove this line if not used.

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java [hunk 9]
```diff
@@ -147,6 +137,7 @@ public void deliveryComplete(IMqttDeliveryToken token) {
                     }
                 });
                 client.connect(options);
+                Thread.sleep(1000);
```
Comment (aloyszhang) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/reader/MqttReader.java:140]: what this `sleep` for?

**Comment by qy-liuhuo** [comment]
I have fixed these issues, please review again, thank you!

**Review by aloyszhang — APPROVED** [review]

**Comment by dockerzhang** [comment]
> I have fixed these issues, please review again, thank you!

please click `resolve` if you fix the comment, thanks.

File: inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java [hunk 10]
```diff
@@ -54,60 +50,76 @@ private List<Reader> splitSqlJob(String topics, String instanceId) {
...
     @Override
     protected void initSource(InstanceProfile profile) {
-
+        try {
+            LOGGER.info("MqttSource init: {}", profile.toJsonStr());
+            topic = profile.get(TaskConstants.TASK_MQTT_TOPIC);
+            mqttReader = new MqttReader(topic);
+            mqttReader.init(profile);
+        } catch (Exception e) {
+            stopRunning();
```
Comment (justinwwhuang) [inline inlong-agent/agent-plugins/src/main/java/org/apache/inlong/agent/plugin/sources/MqttSource.java:83]: Suggest that the exception handling here be consistent with other sources

File: inlong-agent/agent-plugins/src/test/java/org/apache/inlong/agent/plugin/sources/TestMqttReader.java [hunk 11]
```diff
@@ -106,13 +107,13 @@ public void setUp() throws Exception {
 
         when(jobProfile.get(eq(CommonConstants.PROXY_INLONG_GROUP_ID), anyString())).thenReturn(groupId);
```
Comment (justinwwhuang) [inline inlong-agent/agent-plugins/src/test/java/org/apache/inlong/agent/plugin/sources/TestMqttReader.java:108]: Sorry, it is recommended to put the reader's logic in the source. The current reader needs to be deleted after it

Comment (qy-liuhuo) [reply inlong-agent/agent-plugins/src/test/java/org/apache/inlong/agent/plugin/sources/TestMqttReader.java:108]: sorry, do you mean write the MqttReader's code in the mqttsource file? But how to handle the split function, just rerturn null as other sources?

Comment (justinwwhuang) [reply inlong-agent/agent-plugins/src/test/java/org/apache/inlong/agent/plugin/sources/TestMqttReader.java:108]: > sorry, do you mean write the MqttReader's code in the mqttsource file? But how to handle the split function, just rerturn null as other sources?

yes

Comment (qy-liuhuo) [reply inlong-agent/agent-plugins/src/test/java/org/apache/inlong/agent/plugin/sources/TestMqttReader.java:108]: ok, i will fix it

**Review by aloyszhang — APPROVED** [review]

**Review by justinwwhuang — APPROVED** [review]
LGTM

**Pull request merged**